enhancements/network/on-prem-mutable-vips.md

```diff
@@ -0,0 +1,213 @@
+enablement of this feature should be performed with some level of care by the end user.
+
+## Design Details
+
+This feature will require changes in a few different components:
+
+* Installer - New fields will need to be populated during the installation. This is a safe part as installer is already populating `status` and now will populate `status` and `spec`.
+
+* API - New fields will need to be added to the platform spec section of the infrastructure object.
+
+* Cluster Network Operator - New object will need to be watched by one of its controllers. The core logic of validating the requested change will be implemented there.
+
+* Machine Config Operator - Definition of the runtimecfg Pod will need to be rendered again when the VIP configuration changes.
+
+* baremetal-runtimecfg - The code rendering the keepalived template belongs to this repo. As the enhancement does not touch the rendering part no big changes are expected in the runtimecfg. This component is used to re-render configuration in case something changes in the cluster (e.g. new nodes are added) so the fact we modify configuration is not a new feature from its perspective.
+Small changes may be needed to acommodate to the new user scenario to provide good user experience.
+
+### openshift/api
+
+The new structure of platformSpec would look like this:
+
+```yaml
+platformSpec:
+  properties:
+    baremetal:
+      description: BareMetal contains settings specific to the BareMetal platform.
+      type: object
+      properties:
+        apiServerInternalIPs:
+          description: apiServerInternalIPs are the IP addresses to...
+          type: array
+        ingressIPs:
+          description: ingressIPs are the external IPs which...
+          type: array
+      ...
+```
+
+This format (i.e. usage of arrays) is already used by platformStatus so this does not introduce a new type nor schema.
+
+### Cluster Network Operator
+
+There already exists `infrastructureconfig_controller` inside CNO that watches for changes in `configv1.Infrastructure` for on-prem platforms. Because of this we do not need to create a new one nor change any fundaments, we only need to extend its logic.
+
+Currently the controller already reconciles on the `CreateEvent` and `UpdateEvent` so we will implement a set of functions that compare and validate `platformSpec` with `platformStatus` and do what's needed.
+
+### Machine Config Operator
+
+[Keepalived templates managed by the Machine Config Operator](https://github.com/openshift/machine-config-operator/blob/2b31b5b58f0d7d9fe6c3e331e4b3f01c9a1bd00c/templates/common/on-prem/files/keepalived.yaml#L46-L49) currently use the `PlatformStatus` fields. We will need to change the code inside the controller
+which [pulls this value](https://github.com/openshift/machine-config-operator/blob/2b31b5b58f0d7d9fe6c3e331e4b3f01c9a1bd00c/pkg/controller/template/render.go#L551-L575) so that new `PlatformSpec` path is used.
```

Comment:
If CNO is just going to sync spec and status, we probably don't need to change MCO at all. It can just pick up the changes from status when they get set.

Reply:
Of course this is right... I just modified the doc to be clear that MCO observes PlatformStatus today and as CNO will sync Spec to Status, no changes in MCO are needed